Userland/xargs.cpp

```diff
@@ -0,0 +1,186 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/StdLibExtras.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/ArgsParser.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio rpath proc exec", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    const char* command = nullptr;
+    const char* replace_str = "{}";
+    bool split_with_nulls = false;
+    const char* specified_delimiter = "\n";
+    Vector<const char*> arguments;
+    bool verbose = false;
+    const char* file_to_read = "-";
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(command, "Command to run", "command");
+    args_parser.add_positional_argument(arguments, "Initial arguments to pass to command", "initial_args", Core::ArgsParser::Required::No);
+    args_parser.add_option(replace_str, "Value to be replaced in arguments", "replace", 'I', "replace_str");
+    args_parser.add_option(split_with_nulls, "Split input items with the null character instead of newline", "null", '0');
+    args_parser.add_option(specified_delimiter, "Split the input items with the specified character", "delimiter", 'd', "delim");
+    args_parser.add_option(verbose, "Display each command before executing it", "verbose", 'v');
+    args_parser.add_option(file_to_read, "Read arguments from the specified file instead of stdin", "arg-file", 'a', "file");
+    args_parser.parse(argc, argv);
+
+    char entry_separator = split_with_nulls ? '\0' : *specified_delimiter;
+
+    StringView replace_view { replace_str };
+
+    Vector<Vector<StringView>> argument_parts;
+    argument_parts.ensure_capacity(arguments.size());
+    size_t index = 0;
+    bool some_argument_has_replacement = false;
+
+    for (auto argument : arguments) {
+        StringView arg { argument };
+        Vector<StringView> parts;
+        size_t previous_index = 0;
+        if (auto first_index = arg.find_first_of(replace_view); first_index.has_value()) {
+            some_argument_has_replacement = true;
+            while (first_index.has_value()) {
+                auto current_index = first_index.value();
+                auto substring_until_replacement = arg.substring_view(previous_index, current_index);
+                parts.append(substring_until_replacement);
+                previous_index = current_index;
+                first_index = arg.substring_view(current_index + replace_view.length(), arg.length() - current_index - replace_view.length()).find_first_of(replace_view);
```

Comment:
Try one of these:

https://github.com/SerenityOS/serenity/blob/dd08c992e8d884f646bdf9e9e75b2236ad5d38c8/AK/StringView.h#L101-L118

Reply:
And generally, I have troubles understanding what this block of code is doing. Perhaps a comment with some examples would help?

Reply:
That won't cut it, we have
```
foo{}bar
   ^ find_first_of
     ^ where we want to start at
   ^ where substr_after would give for substring_until_replacement
```